contrib/src/main/scala/akka/stream/contrib/SourceGen.scala

```diff
@@ -0,0 +1,120 @@
+ */
+object SourceGen {
+
+  /**
+   * Create a `Source` that will unfold a value of type `S` by
+   * passing it through a flow. The flow should emit a
+   * pair of the next state `S` and output elements of type `E`.
+   * Source completes when the flow completes.
+   */
+  def unfoldFlow[S, E, M](seed: S)(flow: Graph[FlowShape[S, (S, E)], M]): Source[E, M] = {
+
+    val fanOut2Stage = fanOut2unfoldingStage[(S, E), S, E](seed, {
+      (handleStateElement, grabIn, completeStage) =>
+        new InHandler {
+          override def onPush() = {
+            val (s, e) = grabIn()
+            handleStateElement(s, e)
+          }
+        }
+    })
+
+    unfoldFlowGraph(fanOut2Stage, flow)
+  }
+
+  /**
+   * Create a `Source` that will unfold a value of type `S` by
+   * passing it through a flow. The flow should emit an output
+   * value of type `O`, that when fed to the unfolding function,
+   * generates a pair of the next state `S` and output elements of type `E`.
+   */
+  def unfoldFlowWith[E, S, O, M](seed: S, flow: Graph[FlowShape[S, O], M])(unfoldWith: O => Option[(S, E)]): Source[E, M] = {
+
+    val fanOut2Stage = fanOut2unfoldingStage[O, S, E](seed, {
+      (handleStateElement, grabIn, completeStage) =>
+        new InHandler {
+          override def onPush() = {
+            val o = grabIn()
+            unfoldWith(o) match {
+              case None         => completeStage()
+              case Some((s, e)) => handleStateElement(s, e)
+            }
+          }
+        }
+    })
+
+    unfoldFlowGraph(fanOut2Stage, flow)
+  }
+
+  private[akka] def unfoldFlowGraph[E, S, O, M](
+    fanOut2Stage: GraphStage[FanOutShape2[O, S, E]],
+    flow:         Graph[FlowShape[S, O], M]
+  ): Source[E, M] = Source.fromGraph(GraphDSL.create(flow) {
+    implicit b =>
+      {
+        f =>
+          {
+            import GraphDSL.Implicits._
+
+            val fo2 = b.add(fanOut2Stage)
+            fo2.out0 ~> f ~> fo2.in
+            SourceShape(fo2.out1)
+          }
+      }
+  })
+
+  private[akka] def fanOut2unfoldingStage[O, S, E](seed: S, withInHandler: ((S, E) => Unit, () => O, () => Unit) => InHandler) = new GraphStage[FanOutShape2[O, S, E]] {
+
+    override val shape = new FanOutShape2[O, S, E]("unfoldFlow")
+    override def createLogic(attributes: Attributes) = {
+
+      new GraphStageLogic(shape) {
+
+        import shape._
+
+        var pending: S = seed
+        var pushedToCycle = false
+
+        setHandler(in, withInHandler((s, e) => {
+          pending = s
+          push(out1, e)
+          pushedToCycle = false
+        }, () => grab(in), completeStage))
+
+        setHandler(out0, new OutHandler {
+          override def onPull() = if (!pushedToCycle && isAvailable(out1)) {
+            push(out0, pending)
+            pending = null.asInstanceOf[S]
+            pushedToCycle = true
+          }
+
+          override def onDownstreamFinish() = {
+            //Do Nothing, intercept completion as downstream
```

Comment:
The scenario I was thinking about was:
 - no element in cycle
 - feedback becomes unavailable
 - output is pulled -> inlet is pulled
 - inlet does NOT become unavailable

I.e. the inner Flow only cancels. We can still cleanup in this case since we know that there is no element in the cycle so we will never be able to satisfy the downstream pull anyway, so we can just complete.